### PR TITLE
esbuild: fix resolving of deep modules

### PR DESCRIPTION
**Pull request by tlhunter**
### What does this PR do?
- fixes the ESBuild plugin to properly load deep instrumented package modules
- deletes DCITM, moves functionality to `packages/datadog-instrumentations/src/helpers/bundler-register.js`
- uses a single DC channel for all modules instead of a channel per module

### Motivation
- currently it only supports the highest level entry into a package, e.g. `require('express')`
- more deeply nested modules, like most of them we define via `addHook()` with a files arg, were overlooked
- in practice most things don't actually work -- save for the Express integration test 🤯

### CODE REVIEW

File: packages/datadog-instrumentations/src/helpers/hooks.js [hunk 1]
```diff
@@ -38,6 +38,7 @@ module.exports = {
   'find-my-way': () => require('../find-my-way'),
   'fs': () => require('../fs'),
   'node:fs': () => require('../fs'),
+  'generic-pool': () => require('../generic-pool'),
```
Comment (tlhunter) [inline packages/datadog-instrumentations/src/helpers/hooks.js:41]: This was the only instrumentation missing from the list.

File: packages/datadog-esbuild/index.js [hunk 2]
```diff
@@ -2,73 +2,111 @@
...
-const rawBuiltins = require('module').builtinModules
+const path = require('path')
+const fs = require('fs')
+const instrumentations = require('../datadog-instrumentations/src/helpers/instrumentations.js')
 
 warnIfUnsupported()
 
+{
+  // run each instrumentation so that `instrumentations` gets populated
+  const INS_PATH = path.join(__dirname, '../datadog-instrumentations/src')
```
Comment (rochdev) [inline packages/datadog-esbuild/index.js:13]: It would be simpler to just require instead of manipulating the filesystem directly. It would also avoid issues if the structure of the folder changes or if unrelated files are added.

File: packages/datadog-esbuild/index.js [hunk 3]
```diff
@@ -2,73 +2,111 @@
...
+    require(path.join(INS_PATH, entry))
+  }
+}
+
+const modulesOfInterest = new Set()
+
+for (const instrumentation of Object.values(instrumentations)) {
+  for (const entry of instrumentation) {
+    if (!entry.file) {
+      modulesOfInterest.add(entry.name) // redis
```
Comment (rochdev) [inline packages/datadog-esbuild/index.js:26]: This comment is confusing. It took me a while to understand that this has nothing to do with redis.

File: packages/datadog-instrumentations/src/helpers/register.js [hunk 4]
```diff
@@ -19,6 +21,51 @@ const disabledInstrumentations = new Set(
 
 const loadChannel = channel('dd-trace:instrumentation:load')
 
+if (!dc.subscribe) {
```
Comment (rochdev) [inline packages/datadog-instrumentations/src/helpers/register.js:24]: Why not use the API available from the polyfill (and also use the polyfill).

File: packages/datadog-instrumentations/src/helpers/register.js [hunk 5]
```diff
@@ -19,6 +21,51 @@ const disabledInstrumentations = new Set(
...
+}
+if (!dc.unsubscribe) {
+  dc.unsubscribe = (channel, cb) => {
+    if (dc.channel(channel).hasSubscribers) {
+      dc.channel(channel).unsubscribe(cb)
+    }
+  }
+}
+
+dc.subscribe('dd-trace-esbuild', (payload) => {
```
Comment (rochdev) [inline packages/datadog-instrumentations/src/helpers/register.js:37]: Is this really specific to esbuild? Also, why not have this in another file like before?

File: packages/datadog-esbuild/index.js [hunk 6]
```diff
@@ -80,23 +118,26 @@ module.exports.setup = function (build) {
...
+      const ch = dc.channel('dd-trace-esbuild');
+      const mod = require('${path}');
       const payload = {
         module: mod,
-        path: ${JSON.stringify(args.path)},
-        version: ${JSON.stringify(args.pluginData.version)}
+        path: '${data.raw}',
+        version: '${data.version}',
+        package: '${data.pkg}',
+        relPath: '${data.path}'
```
Comment (rochdev) [inline packages/datadog-esbuild/index.js:135]: Is the distinction between `path` and `relPath` actually needed?

File: packages/datadog-esbuild/index.js (hunk 2 above)
Comment (tlhunter) [reply packages/datadog-esbuild/index.js:13]: Hmm... I suppose I can require `packages/datadog-instrumentations/src/helpers/hooks.js` then loop through and execute each object value.

Comment (tlhunter) [reply packages/datadog-esbuild/index.js]: This change has been made.

File: packages/datadog-esbuild/index.js (hunk 3 above)
Comment (tlhunter) [reply packages/datadog-esbuild/index.js]: I fixed up the two comments, they should be more self-explanatory now.

File: packages/datadog-instrumentations/src/helpers/register.js (hunk 5 above)
Comment (tlhunter) [reply packages/datadog-instrumentations/src/helpers/register.js]: I suppose it's more specific to bundling. Went with `dd-trace:bundler:load`.

File: packages/datadog-esbuild/index.js (hunk 6 above)
Comment (tlhunter) [reply packages/datadog-esbuild/index.js]: I modified the code so that the emitter makes the decision of how to handle raw != package and now one less piece of data is being emitted.

File: packages/datadog-instrumentations/src/helpers/register.js (hunk 4 above)
Comment (tlhunter) [reply packages/datadog-instrumentations/src/helpers/register.js]: This was taken mostly as-is from DCITM. If you recall from your DC refactor it's the only occasion where we still use the built-in DC instead of the polyfill. The issue is that we can't require dd-trace modules directly from within the bundled interpositional code.

File: packages/datadog-instrumentations/src/helpers/register.js (hunk 5 above)
Comment (tlhunter) [reply packages/datadog-instrumentations/src/helpers/register.js]: Also, this has been moved to a new file.
